v-next/core/src/internal/providers/http.ts

```diff
@@ -0,0 +1,361 @@
+  parseJsonRpcResponse,
+} from "../utils/json-rpc.js";
+import { getHardhatVersion } from "../utils/package.js";
+
+import { ProviderError, ProviderErrorCode } from "./errors.js";
+
+const TOO_MANY_REQUEST_STATUS = 429;
+const MAX_RETRIES = 6;
+const MAX_RETRY_WAIT_TIME_SECONDS = 5;
+
+export class HttpProvider extends EventEmitter implements EIP1193Provider {
+  readonly #url: string;
+  readonly #networkName: string;
+  readonly #extraHeaders: Record<string, string>;
+  readonly #dispatcher: Dispatcher;
+  #nextRequestId = 1;
+
+  /**
+   * Creates a new instance of `HttpProvider`.
+   */
+  public static async create({
+    url,
+    networkName,
+    extraHeaders = {},
+    timeout,
+  }: {
+    url: string;
+    networkName: string;
+    extraHeaders?: Record<string, string>;
+    timeout: number;
+  }): Promise<HttpProvider> {
+    if (!isValidUrl(url)) {
+      throw new HardhatError(HardhatError.ERRORS.NETWORK.INVALID_URL, {
+        value: url,
+      });
+    }
+
+    const dispatcher = await getHttpDispatcher(url, timeout);
+
+    const httpProvider = new HttpProvider(
+      url,
+      networkName,
+      extraHeaders,
+      dispatcher,
+    );
+
+    return httpProvider;
+  }
+
+  /**
+   * @private
+   *
+   * This constructor is intended for internal use only.
+   * Use the static method {@link HttpProvider.create} to create an instance of
+   * `HttpProvider`.
+   */
+  constructor(
+    url: string,
+    networkName: string,
+    extraHeaders: Record<string, string>,
+    dispatcher: Dispatcher,
+  ) {
+    super();
+
+    this.#url = url;
+    this.#networkName = networkName;
+    this.#extraHeaders = extraHeaders;
+    this.#dispatcher = dispatcher;
+  }
+
+  /**
+   * Sends a JSON-RPC request.
+   *
+   * @param requestArguments The arguments for the request. The first argument
+   * should be a string representing the method name, and the second argument
+   * should be an array containing the parameters. See {@link RequestArguments}.
+   * @returns The `result` property of the successful JSON-RPC response.
+   * @throws {ProviderError} If the JSON-RPC response indicates a failure.
+   * @throws {HardhatError} with descriptor:
+   * - {@link HardhatError.ERRORS.NETWORK.INVALID_REQUEST_PARAMS} if the
+   * params are not an array.
+   * - {@link HardhatError.ERRORS.NETWORK.CONNECTION_REFUSED} if the
+   * connection is refused.
+   * - {@link HardhatError.ERRORS.NETWORK.NETWORK_TIMEOUT} if the request
+   * times out.
+   */
+  public async request(
+    requestArguments: RequestArguments,
+  ): Promise<SuccessfulJsonRpcResponse["result"]> {
+    const { method, params } = requestArguments;
+
+    const jsonRpcRequest = getJsonRpcRequest(
+      this.#nextRequestId++,
+      method,
+      params,
+    );
+    const jsonRpcResponse = await this.#fetchJsonRpcResponse(jsonRpcRequest);
+
+    if (isFailedJsonRpcResponse(jsonRpcResponse)) {
+      const error = new ProviderError(jsonRpcResponse.error.code);
```

Comment:
See https://github.com/NomicFoundation/hardhat/pull/5580/files#r1715725985

Reply:
this also makes me feel that there's some test missing somewhere

Reply:
> this also makes me feel that there's some test missing somewhere
I believe this doesn't apply anymore after the refactor, but let me know if you think otherwise.